internal/handler/callback/provisionwatcher.go

```diff
@@ -0,0 +1,115 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package callback
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/edgexfoundry/device-sdk-go/internal/cache"
+	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/internal/provision"
+	"github.com/google/uuid"
+)
+
+func handleProvisionWatcher(method string, id string) common.AppError {
+	ctx := context.WithValue(context.Background(), common.CorrelationHeader, uuid.New().String())
+	switch method {
+	case http.MethodPost:
+		handlePostCallback(ctx, id)
+	case http.MethodPut:
+		handlePutCallback(ctx, id)
+	case http.MethodDelete:
+		handleDeleteCallback(id)
+	default:
+		common.LoggingClient.Error(fmt.Sprintf("Invalid provisionwatcher method type: %s", method))
+		appErr := common.NewBadRequestError("Invalid provisionwatcher method", nil)
+		return appErr
+	}
+
+	return nil
+}
+
+func handlePostCallback(ctx context.Context, id string) common.AppError {
+	pw, err := common.ProvisionWatcherClient.ProvisionWatcher(id, ctx)
+	if err != nil {
+		appErr := common.NewBadRequestError(err.Error(), err)
+		common.LoggingClient.Error(fmt.Sprintf("Cannot find provisionwatcher %s in Core Metadata: %v", id, err))
+		return appErr
+	}
+
+	_, exist := cache.Profiles().ForName(pw.Profile.Name)
+	if !exist {
+		err = cache.Profiles().Add(pw.Profile)
+		if err == nil {
+			provision.CreateDescriptorsFromProfile(&pw.Profile)
+			common.LoggingClient.Info(fmt.Sprintf("Added device profile %s", pw.Profile.Name))
+		} else {
+			appErr := common.NewServerError(err.Error(), err)
+			common.LoggingClient.Error(fmt.Sprintf("Cannot add device profile %s: %v", pw.Profile.Name, err.Error()))
+			return appErr
+		}
+	}
+
+	err = cache.ProvisionWatchers().Add(pw)
+	if err == nil {
+		common.LoggingClient.Info(fmt.Sprintf("Added provisionwatcher %s", id))
+	} else {
+		appErr := common.NewServerError(err.Error(), err)
+		common.LoggingClient.Error(fmt.Sprintf("Cannot add provisionwatcher %s: %v", id, err.Error()))
+		return appErr
+	}
+
+	return nil
+}
+
+func handlePutCallback(ctx context.Context, id string) common.AppError {
+	pw, err := common.ProvisionWatcherClient.ProvisionWatcher(id, ctx)
+	if err != nil {
+		appErr := common.NewBadRequestError(err.Error(), err)
+		common.LoggingClient.Error(fmt.Sprintf("Cannot find provisionwatcher %s in Core Metadata: %v", id, err))
+		return appErr
+	}
+
+	_, exist := cache.Profiles().ForName(pw.Profile.Name)
+	if !exist {
+		err = cache.Profiles().Add(pw.Profile)
+		if err == nil {
+			provision.CreateDescriptorsFromProfile(&pw.Profile)
```

Comment:
we don't need to manage ValueDescriptor here, because it will be handled in Device callback